prime-router/src/main/kotlin/transport/SoapTransport.kt

```diff
@@ -0,0 +1,237 @@
+        // get the external file name to send to the client, if we need it
+        val fileName = header.reportFile.externalName
+        context.logger.info(
+            "Preparing to send ${header.reportFile.reportId} " +
+                "to ${soapTransportType.soapAction} at ${soapTransportType.endpoint}"
+        )
+        // based on who we are sending this report to, we need to get the credentials, and we also need
+        // to create the actual implementation object based on who we're sending to
+        val credential = lookupCredentials(receiver)
+        // calling the SOAP serializer in an attempt to be somewhat abstract here. this is based on the
+        // namespaces provided to the SOAP transport info. It's kind of BS magical string garbage, but the
+        // reality is that each client could have different objects we have to create for them, and these
+        // objects might nest, and/or have different constructors, and it's really hard to be really generic
+        // about this kind of stuff. So this is some semi-tight coupling we will just have to manage.
+        // And honestly, if a client changes their SOAP endpoint, we'd (probably) need to do some coding
+        // on our end, so incurring the maintenance cost here is (probably) okay.
+        val xmlObject = SoapObjectService.getXmlObjectForAction(soapTransportType, header, context, credential)
+            ?: error("Unable to find a SOAP object for the namespaces provided")
+        // wrap the object in the generic envelope. At least this gets to be generic
+        val soapEnvelope = SoapEnvelope(xmlObject, soapTransportType.namespaces ?: emptyMap())
+
+        return try {
+            // run our call to the endpoint in a blocking fashion
+            runBlocking {
+                launch {
+                    val responseBody = connectToSoapService(
+                        soapEnvelope.toXml(),
+                        soapTransportType.endpoint,
+                        soapTransportType.soapAction,
+                        context,
+                        httpClient ?: createDefaultHttpClient()
+                    )
+                    // update the action history
+                    val msg = "Success: SOAP transport of $fileName to $soapTransportType:\n$responseBody"
+                    context.logger.info("Message successfully sent!")
+                    actionHistory.trackActionResult(msg)
+                    actionHistory.trackSentReport(
+                        receiver,
+                        sentReportId,
+                        fileName,
+                        soapTransportType.toString(),
+                        msg,
+                        header.reportFile.itemCount
+                    )
+                    actionHistory.trackItemLineages(Report.createItemLineagesFromDb(header, sentReportId))
+                }
+            }
+            // return null
+            null
+        } catch (t: Throwable) {
```

Comment:
We should not treat all errors as something for retry and warnings. 4XX returned by the server is always a client error if I read the SOAP specs correctly. We should log errors. They should trigger a pager alert. This code just warns. 

With SOAP, I believe that an HTTP 500 is returned with all SOAP faults. SOAP faults can either be server errors (which we should retry) or client errors (which we should not retry). So I think we need code to examine the SOAP response.

Reply:
@RickHawesUSDS Would something like this work for now?
```kotlin
// If Ktor fails to connect, or the server returns an error code, it is thrown
// as an exception higher up, which we catch and then track here. We do not need
// to worry about capturing and parsing out the return value from the response
// because Ktor treats errors as exceptions
val msg = "FAILED SOAP of inputReportId ${header.reportFile.reportId} to " +
    "$soapTransportType (orgService = ${header.receiver.fullName})" +
    ", Exception: ${t.localizedMessage}"
context.logger.severe(msg)
context.logger.severe(t.stackTraceToString())
// do some additional handling of the error here. if we are dealing with a 400 error, we
// probably don't want to retry, and we need to stop now
// if the error is a 500 we can do a retry, but both should probably throw a pager duty notification
when (t) {
    is ClientRequestException -> {
        (t as ClientRequestException).let {
            context.logger.severe(
                "Received ${it.response.status.value}: ${it.response.status.description} " +
                    "requesting ${it.response.request.url}. This is not recoverable. Will not retry."
            )
        }
        actionHistory.setActionType(TaskAction.send_error)
        actionHistory.trackActionResult(msg)
        null
    }
    is ServerResponseException -> {
        // this is largely duplicated code as below, but we may want to add additional
        // instrumentation based on the specific error type we're getting. One benefit
        // we can use now is getting the specific response information from the
        // ServerResponseException
        (t as ServerResponseException).let {
            context.logger.severe(
                "Received ${it.response.status.value}: ${it.response.status.description} " +
                    "from server ${it.response.request.url}. This may be recoverable. Will retry."
            )
        }
        actionHistory.setActionType(TaskAction.send_warning)
        actionHistory.trackActionResult(msg)
        RetryToken.allItems
    }
    else -> {
        // this is an unknown exception, and maybe not one related to ktor, so we should
        // track, but try again
        actionHistory.setActionType(TaskAction.send_warning)
        actionHistory.trackActionResult(msg)
        RetryToken.allItems
    }
}
```